####  Dependencies Required

In [24]:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler

#### Using 2019 Data to Train the Model and Test Against 2020 Data

In [25]:
train_df = pd.read_csv(Path('2019loans.csv'))
test_df = pd.read_csv(Path('2020Q1loans.csv'))

In [26]:
train_df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0     7000.0    0.1894       256.38       MORTGAGE     75000.0   
1    40000.0    0.1614       975.71       MORTGAGE    102000.0   
2    11000.0    0.2055       294.81           RENT     45000.0   
3     4000.0    0.1612       140.87       MORTGAGE     38000.0   
4    14000.0    0.1797       505.93       MORTGAGE     43000.0   

  verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0        Not Verified          n  28.62          0.0             2.0  ...   
1     Source Verified          n  11.72          2.0             0.0  ...   
2            Verified          n  37.25          1.0             3.0  ...   
3        Not Verified          n  42.89          1.0             0.0  ...   
4     Source Verified          n  22.16          1.0             0.0  ...   

   percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0              87.5                   0.0        0.0         352260.0   
1               0.0                   0.0        0.0         294664.0   
2               7.7                   0.0        0.0          92228.0   
3             100.0                   0.0        0.0         284273.0   
4              25.0                   0.0        0.0         120280.0   

  total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  \
0           62666.0         35000.0                     10000.0   
1          109911.0          9000.0                     71044.0   
2           36007.0         33000.0                     46328.0   
3           52236.0         13500.0                     52017.0   
4           88147.0         33300.0                     78680.0   

   hardship_flag  debt_settlement_flag    target  
0              N                     N  low_risk  
1              N                     N  low_risk  
2              N                     N  low_risk  
3              N                     N  low_risk  
4              N                     N  low_risk  

[5 rows x 84 columns]

#### Assinging X & y train data

In [27]:
X_train = pd.get_dummies(train_df.drop(columns=['target']))
y_train = train_df['target']

In [28]:
X_train.head()

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0     7000.0    0.1894       256.38     75000.0  28.62          0.0   
1    40000.0    0.1614       975.71    102000.0  11.72          2.0   
2    11000.0    0.2055       294.81     45000.0  37.25          1.0   
3     4000.0    0.1612       140.87     38000.0  42.89          1.0   
4    14000.0    0.1797       505.93     43000.0  22.16          1.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0             2.0      20.0      0.0    40414.0  ...   
1             0.0      10.0      0.0    43531.0  ...   
2             3.0      23.0      0.0     8242.0  ...   
3             0.0       7.0      0.0    12767.0  ...   
4             0.0      22.0      0.0    11182.0  ...   

   verification_status_Verified  pymnt_plan_n  initial_list_status_f  \
0                             0             1                      1   
1                             0             1                      0   
2                             1             1                      0   
3                             0             1                      0   
4                             0             1                      0   

   initial_list_status_w  application_type_Individual  \
0                      0                            1   
1                      1                            1   
2                      1                            1   
3                      1                            0   
4                      1                            1   

   application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                           0                1                0   
1                           0                1                0   
2                           0                1                0   
3                           1                1                0   
4                           0                1                0   

   debt_settlement_flag_N  debt_settlement_flag_Y  
0                       1                       0  
1                       1                       0  
2                       1                       0  
3                       1                       0  
4                       1                       0  

[5 rows x 92 columns]

#### Assigning X & y test data

In [29]:
X_test = pd.get_dummies(test_df.drop(columns=['target']))
y_test = test_df['target']

#### Adding Missing Dummy Variables

In [30]:
for col in X_train.columns:
    if col not in X_test.columns:
        X_test[col] = 0

### Logistic Regression

In [31]:
clf = LogisticRegression(solver='lbfgs', random_state=1)
clf.fit(X_train, y_train)
clf.score(X_test, y_test)

C:\Users\drewp\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.5074436410038281

#### Logistic Regression is only 50.7% accurate

### Random Forest Classification Model

In [32]:
clf = RandomForestClassifier()
clf.fit(X_train, y_train)
clf.score(X_test, y_test)

0.6427052318162484

#### Random Forest Classifier is 64.2% accurate. This is a more accurate model thank logistic Regression.

#### Scaling Data

In [33]:
scaler = StandardScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

### Logistic Regression Model with Scaled Data

In [34]:
clf = LogisticRegression(solver='lbfgs')
clf.fit(X_train_scaled, y_train)
clf.score(X_test_scaled, y_test)

C:\Users\drewp\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.7598894087622289

#### The Logistic Regression Model with Scaled Data improved to 75.98% Accuracy

### Random Forest Classifier with Scaled Data

In [35]:
clf = RandomForestClassifier()
clf.fit(X_train_scaled, y_train)
clf.score(X_test_scaled, y_test)

0.6399404508719694

#### Random Forest Classificaton model did not improve when scaling the data. Only the Logistic Regression model was affected when scaling the data.